Rename and rework File::Stat as File::Info

RX14 opened the pull request:
I've aimed to make `File::Info` easier to use, more portable, and nicer than `File::Stat`.

straight-shoota commented on src/crystal/system/unix/file.cr:
```diff
@@ -70,26 +70,65 @@ module Crystal::System::File
+    flags |= ::File::Flags::SetGroup if (stat.st_mode & LibC::S_ISGID) != 0
+    flags |= ::File::Flags::Sticky if (stat.st_mode & LibC::S_ISVTX) != 0
+
+    modification_time = ::Time.new(stat.st_mtim, ::Time::Kind::Utc)
```
you should better use `::Time.utc`

RX14 replied:
`st_mtim` is a `LibC::Timespec`. You can't use `Time.utc`.

straight-shoota replied:
ah, nvm

lbguilherme commented on src/file.cr:
```diff
@@ -16,7 +16,7 @@ class File < IO::FileDescriptor
 
   # :nodoc:
-  DEFAULT_CREATE_MODE = LibC::S_IRUSR | LibC::S_IWUSR | LibC::S_IRGRP | LibC::S_IROTH
+  DEFAULT_CREATE_PERMISSIONS = 0o644
```
Maybe it is better to use `File::Permissions.flags(OwnerRead, OwnerWrite, GroupRead, OtherRead)` here?

RX14 replied:
I don't think so, the shorter octal syntax is far easier to read. "everyone" knows unix octal permission syntax.

ysbaddaden replied:
Then why bother with a `File::Permissions` enum?

asterite commented on src/file/info.cr:
```diff
@@ -0,0 +1,107 @@
+  # The binary representation of this enum is defined to be same representation
+  # as the permission bits of a unix `st_mode` field. `File::Permissions`
+  # can also be compared to it's underlying bitset, for example
+  # `File::Permissions::All == 0o777` will always be `true`.
```
What about the opposite? Does `0o777 == File::Permissions::All` work?

asterite replied:
Also, maybe `always` isn't neccessary here.

RX14 replied:
Ah, yeah. Perhaps this equality was a bad idea. Or should I just reopen `struct Int`? It's a bit ugly how these equalities are defined in two places.

RX14 commented:
@asterite yeah, while debugging I found that I really wanted to be able to see the permissions better so I implemented that.

ysbaddaden commented on src/file/info.cr:
```diff
@@ -0,0 +1,107 @@
+
+    # :nodoc:
+    def initialize(@size, @permissions, @type, @flags, @modification_time, @owner, @group)
+    end
```
It lost query methods, such as `symlink?`, `file?`, ... which is an unwelcomed change.

RX14 replied:
It's just `type.file?` now.

straight-shoota replied:
We could think about forwarding these methods to `type` similar to the weekday names in `Time`. Allthough this would effectively be aliases. But the API would be simpler that way. You shouldn't have to write `type` in between to ask a `File::Info` if it describes a symlink.

RX14 replied:
Why not? It's really just some extra typing and it means you have a clear separation between permissions, types, and flags on files. Which I think is really nice.

straight-shoota replied:
I see your argument. But I think that's too bulky.

Sija replied:
Why not use `delegate` here?

RX14 commented:
@ysbaddaden Information has been intentionally left out to make it portable. `atime` is disabled on most modern distros, either with `relatime` or `noatime` mount options (and nobody uses it) and `ctime` is usually not what people want or think it is (they want `btime` which basically no filesystems support). `mtime` is really the only widely supported and actually useful time. Go only provides modification time too.

I'll benchmark old stat vs new stat but i'm not sure that a bit of bit twiddling is going to make much difference when put next to the price of a syscall, and making it wrap `LibC::Stat` would move platform-specifics into the definition of `File::Info` instead of just the construction. Keep in mind that on windows I'm not going to be using `LibC::Stat` at all, but recostructing the same information from `GetFileInformationByHandle`.

RX14 commented:
Benchmarks say it doesn't matter *much*. I'd definitely say it's not worth optimizing for a while.

old:
```
 File.stat   1.77M (565.92ns) (± 4.62%)  0 B/op   1.77× slower
File.lstat   1.76M (568.21ns) (± 4.58%)  0 B/op   1.78× slower
 File#stat   3.13M (319.46ns) (± 3.19%)  0 B/op        fastest
```

new:
```
 File.stat   1.74M (575.88ns) (± 4.60%)  0 B/op   1.79× slower
File.lstat   1.69M (590.26ns) (±10.11%)  0 B/op   1.84× slower
 File#stat   3.11M (321.26ns) (± 3.22%)  0 B/op        fastest
```

RX14 commented:
Also keep in mind that benchmark is a best-case for stat: one file only in a loop. When you add in a typical usage where the stat may not be in cache all the time you're going to see far less than the 2-4% overhead here.

bew commented on spec/std/file_spec.cr:
```diff
@@ -325,20 +325,31 @@ describe "File" do
+    it "can take File::Permissions" do
+      path = "#{__DIR__}/data/chmod.txt"
+      begin
+        File.write(path, "")
```
oh ok, probably not worse to change it here then

RX14 commented:
I do wonder whether we should rename `File.stat?` as `File.info?` and have `File.info?(follow_symlink: false)`.

Also i'm probably going to just rename `File::Type::SymbolicLink` to `File::Type::Symlink`, or add an alias.

asterite commented:
I actually learned in this PR that `lstat` doesn't (or does) follow symlinks, so a named argument to select that behavior is better in my opinion. `lstat` is too cryptic. So maybe it's better to just use `info`.

I believe Ruby started with "works only in unix" and eventually had to keep all those cryptic, unix-specific names. Go for example doesn't do this, it provides a clear interface to the computer without using ancient strange names. I think we should do the same, even if it means a bigger barrier to those coming from Ruby.

RX14 commented:
@asterite its called stat and lstat on go too...

asterite commented:
Maybe because go isn't that flexible with method names.

RX14 commented:
@asterite I do agree it should be `File.info`. (eventually `Path#info` and `File#info` for stat/lstat and fstat respectively)

RX14 commented:
Updated this PR to take into account the above suggestions.

RX14 commented:
Looks like I accidentally made this PR depend on https://github.com/crystal-lang/crystal/pull/5619...

RX14 commented:
Pushed again to fix this, this can be re-reviewed now.

ysbaddaden commented on src/file/info.cr:
```diff
@@ -0,0 +1,117 @@
+end
+
+struct Int
+  def ==(other : File::Permissions)
```
I'm not sure reopening `Int` is a good idea.

RX14 replied:
And the alternative is?

ysbaddaden replied:
Not support `Int==(File::Permissions)` nor `File::Info#permissions==(Int)`. You must wrap the octal value. I believe it's better than reopening Int for a corner case —checking *exact* permissions equality doesn't seem like a usual case.

RX14 commented:
> Why change stat naming to info? it's still an abbreviation and doesn't tell any better.

`info` implies you're getting additional data, or metadata, about something. `stat` (if you can decode what it means at all) implies to me statistics about the fd/filename, which is absolutely not the case.

> Why decode & copy everything (check each file type, decode/copy permissions, generate Time instances) instead of using a wrapper for on-demand generation? Sure File::Stat would be a mere delegation struct for the underlying POSIX vs Windows implementations, mostly to hold documentation, but the optimizer will inline calls, or we can just flag them with @[AlwaysInline].

Because it's shorter code (look at the length of `src/file/info.cr` vs `src/file/stat.cr`) and easier to read and all the platform-specifics are nicely contained in a single method in `src/crystal/system`. And I benchmarked it to show that the performance impact was negligible.

> Having to type .info.type.directory? over .stat.directory? can feel minor, but that's still an unwelcomed hick; forward calls from the struct to the enum, please.

There's no point. `fd.info.type.directory?` can never return true since we don't expose `info` on `opendir` FDs, and every other time you can just use `Dir.exists?` or `File.directory?`. If you're trying to get multiple bits of info out of a single stat call, your code length is hardly going to be impacted hugly by 5 extra characters.

> ctime & atime are still missing (!)

They're not useful, and actively misleading, as I explained before:

> Information has been intentionally left out to make it portable. `atime` is disabled on most modern distros and OSes, either with `relatime` or `noatime` mount options (they're disabled because nobody uses it) and `ctime` is usually not what people want or think it is (they want `btime` which basically no filesystems support). `mtime` is really the only widely supported and actually useful time. Go only provides modification time.

j8r commented:
It would be too bad to remove the `ctime` feature: sometimes we want to know when a file has been renamed or moved. Why not having a `File.info(attributes: false)`? Maybe a `access: false` for `atime`, not sure.

RX14 commented:
Would be cool to get this discussed and finalised since it's blocking windows.

asterite commented:
Just to add a bit more to this, the names and API were copied from Go. I think "info" is better than "stat" as it represents information, not statistics (well, it could be "status", but I don't know).

I also searched other languages. For example in Elixir it's called [File.Stat](https://hexdocs.pm/elixir/File.Stat.html) but right in the beginning of those docs it says "A struct that holds file **information**. In Erlang, this struct is represented by a **:file_info** record". So why did they chose "Stat" for a name? :-P

Then we have [Julia](https://docs.julialang.org/en/release-0.4/stdlib/file/#Base.stat): it's called "stat", but "contain information about the file".

I don't know about the removal of "ctime" and "atime", but since Go doesn't provide that I guess that's fine (Go developers know more than us).

RX14 commented:
To be clear - I'm not super invested in removing ctime and atime. I just started from the interface that Go provides and until now nobody's given me a usecase for atime and ctime.

straight-shoota commented:
I'd say atime and ctime should not really the issue here. I don't see a usecase for them either, but if one emerges, they can easily be re-added later. The same goes for abbreviations for `file.type`. Let's not bother with these details that don't matter much to the general implementation.

Info instead of stat is fine, I guess. An experienced developer might be more likely looking for "stat" but will get around with info wich is IMHO a better name that's more descriptive and easier to understand without any background.

lbguilherme commented:
My argument against `stat` and for `info` is that stat is a POSIX naming, and we are going to support non-POSIX operating systems (Windows), better have a more agnostic stdlib, so it feels native for everybody.

j8r commented:
A use case for `ctime`: print the last time the file has been moved – can be very useful to know the real date since a file is in its directory. Even more with multiple mount points, but don't know if the `mtime` of a file changes if moved from one to another (e.g. a NFS).

All we need is to add a parameter that adds the last change time of attributes.

RX14 commented:
@j8r if we include `atime` and `ctime` we'd include it all the time, not optionally.

My complaint though is that you described what `ctime` is *for* but not an actual usecase. When do you need to know this?

j8r commented:
@RX14 For backups, when used with retention time. For example you keep the archive in a place, then move it to a NFS.

As I said above, why not having `File.info(attributes: false)`? When true, it returns `ctime`, false the `mtime`?

asterite commented:
@j8r it's not clear what you mean when you say `attributes: false`. It returns an object that suddenly has more attributes? This is a compiled language, that's not possible.

j8r commented:
@asterite sorry if I was clear.`ctime` is about the last time when sinc the file **or** its attributes has changed. The boolean `attributes` is about aking into account the file's attributes, a bit like taking into account the file's symlink nature if `follow_symlink: true`. To sumup all the possibilities, with what's called behind:

```crystal
File.info(follow_symlink: false, attributes: false) # => LibC::Stat
File.info(follow_symlink: false)attributes: false) # => LibC::Stat 
File.info(follow_symlink: true, attributes: false) # => LibC::Stat
File.info(follow_symlink: false, attributes: false) # => LibC::Stat
```

j8r commented:
@asterite sorry, `attributes` will be on the `modification_time `method.`ctime` is about the last time since the file **or** its attributes has changed. The boolean `attributes` is about aking into account the file's attributes. `false` will be its default value because this is what most people expect.
```crystal
File.info(myfile).modication_time(attributes: false) #=> mtime
File.info(myfile).modication_time(attributes: true)  #=> ctime
```

asterite commented:
I still don't see why passing `attributes: false` returns the modification time, but `attributes: true` returns the creation time. I see no connection between `attributes` and these two things.

j8r commented:
@asterite `ctime` and `mtime` are explained [here](https://www.quora.com/What-is-the-difference-between-mtime-atime-and-ctime), and is about [file's attributes](https://en.m.wikipedia.org/wiki/File_attribute), that are a type of FS metadata.

RX14 commented:
@ysbaddaden I think you're the only remaining "no", can we merge this?

ysbaddaden commented on src/crystal/system/unix/file.cr:
```diff
@@ -70,26 +70,64 @@ module Crystal::System::File
+    when LibC::S_IFSOCK
+      type = ::File::Type::Socket
+    else
+      raise "BUG: unknown File::Type"
```
It's not a bug, it's an UNKNOWN file type. What if POSIX adds a new file type in the future? We crash?

RX14 commented:
We can't support `ctime` (which is really `metadata_modification_time`) because windows simply doesn't support it. Windows records and saves file creation time [*even over deleting a file and recreating it with the same name*](https://blogs.msdn.microsoft.com/oldnewthing/20050715-14/?p=34923/). I think exposing this complexity to users is insane, so let's just not do it.

I'd be happy to expose `creation_time` as `btime` on unix and `creation_time` on windows, but that requires `statx` which is a more expensive syscall. I suggest that that's another issue.

As for the issue of wrapping or converting structs, I defer that to a vote among the core team.

RX14 commented:
Finally updated this, I think you'll like this much better @ysbaddaden. Now comes with `File::Info#==` for checking if two files are the same (dev and inode).

straight-shoota commented on src/crystal/system/unix/file_info.cr:
```diff
@@ -0,0 +1,60 @@
+
+  def modification_time : ::Time
+    {% if flag?(:darwin) %}
+        ::Time.new(@stat.st_mtimespec, ::Time::Location::UTC)
```
wrong indent

straight-shoota replied:
And using `Time.utc` would be better.

RX14 replied:
If you check the very first review comment on this PR you'll see you've already asked this question.

The indent is wrong though, i'll fix that.

straight-shoota replied:
Damn. It's been to long...

straight-shoota commented:
Looks good.

As a side note, it should also be considered for `File::Info` to work with remote and virtual file systems as well. The suggested implementation should make this possible.

RX14 commented:
@ysbaddaden would be really nice if you could re-review this, since the design of this is blocking more windows work.

RX14 commented:
No-one :(

bararchy commented:
@RX14 can't you and @straight-shoota approve this together ?

RX14 commented:
No? @straight-shoota isn't on the core team. Plus I really want @ysbaddaden's input because he was the original person who made me change my design.

straight-shoota commented:
And it needs a rebase anyway. 😉

RX14 commented:
I've added some shortcuts for `file?` `directory?` and `symlink?`, as they are the common options. I've also fixed `DEFAULT_CREATE_PERMISSIONS`. Looking forward to finally getting this merged!

straight-shoota commented:
I don't like delegating only a few type predicates. It should be either all (my preference) or none to provide the least surprise.

sdogruyol commented:
I prefer having the most common ones than having all / nothing. IMHO this is good to go as it is :+1:

RX14 commented:
@ysbaddaden After looking at https://github.com/crystal-lang/shards/issues/203 I'm wondering if making `File::Info#==` perform samefile was correct. Perhaps it should have been `File.info(foo).same_file? File.info(bar)`? `File.same_file?(foo, bar)`? I'd like other people's input on this API.

ysbaddaden commented:
Maybe just `File.same?(foo, bar)`.

RX14 commented:
I don't think we should expose inode and device numbers (nicely). I'd much rather just expose samefile.

bcardiff commented:
`File.same?(foo, bar)` LGTM, but I would propose something else

Given the following files:

```
$ touch a.txt
$ ln -s a.txt b.txt
$ touch c.txt
```

With the stdlib at master we currently have:

```cr
a = File.new("./a.txt") 
a_copy = File.new("./a.txt") 
```

How to check if a1 and a1_copy refer to the same file?

```cr
a == a_copy # => false
a.info == a_copy.info # => true
```

Ok, but how that works with b?

```cr
b = File.new("./b.txt")
a.info == b.info # => true
```

How can I know that a and b are different files?

```cr
File.info?(a.path, follow_symlinks: false) == File.info?(a_copy.path, follow_symlinks: false) # => true
File.info?(a.path, follow_symlinks: false) == File.info?(b.path, follow_symlinks: false) # => false
```

Ok, but what about if I use only the path of the `File`?
File instances initialized with different path indeed have different path

```cr
a_diff = File.new("a.txt")
a.path      # => "./a.txt"
a_diff.path # => "a.txt"
```

You can use `File.expand_path`, but there are a lot of rules to expand the
path and we had already a File instance

```cr
File.expand_path(a.path)      # => /full/path/to/a.txt
File.expand_path(a_diff.path) # => /full/path/to/a.txt
```

So, what about?

1. Changing `File#info` to `File#info(follow_symlinks = true)`
2. Add `File.same?(file1, file2, follow_symlinks = true)`
3. Remove override of `File::Info==`

That would make simpler all of the above scenarios with `File.same?` having `follow_symlinks` set to one or another value.

RX14 commented:
Because `file.info(follow_symlinks: false)` doesn't make sense. `file.info != File.info(file.path)`. `file.info` is `fstat` and `File.info` is `stat/lstat`.

RX14 commented:
> How can I know that a and b are different files?

They *are* the same file. The whole premise for that question is logically wrong. The confusing part here is actually storing `File#path` in the first place. Not all files even have a (valid) path. You can unlink a path from the filesystem while it is open in a program just fine. And any open programs will keep the inode around in the filesystem, and the file will continue to use up disk space until the program is closed.

bcardiff commented:
Ok, so @RX14 you say to only

1. Add `File.same?(file1, file2)` that will always "follow symlinks" in the comparison
2. Remove override of File::Info==
3. Treat path comparison in another way. Later.

I'm fine with that also. 

Although the premise for the question might seems wrong, but I think it depends on the use case. If someone wants to deal with lower level of file system some api are not the best. But is not the more common use case and it is still doable.

ysbaddaden commented:
Mixing symlinks with hard links to detect file identify is IMHO introducing confusing behavior. Furthermore, detecting hard links is quite specific, and not widely used. We don't need a abstraction for this, hence why I'm proposing to keep exposing inodes on POSIX systems and be done with it.

So:
- detecting hard links: `File.info(a).inode == File.info(b).inode`;
- detecting symlinks: `File.real_path(a) == File.real_path(b)`;
- detecting hard links through potential symlinks:
  `File.info(a, follow_symlinks: true).inode == File.info(b, follow_symlinks: true).inode`.

Simple, does the job, no confusing behavior, and allows finer grained detection. All it needs is to expose a value that we *already* have.

RX14 commented:
@bcardiff having a `follow_symlinks` in `File.same` is fine because symlinks are a property of a path, which `File.same` is working with. It's not fine to have `File#info` with a `follow_symlinks` option because you've already dereferenced the symlink and to to a specific device and inode when you call `open(2)`.

@ysbaddaden the reason why I want to only expose equality only is because your examples are *wrong*! You must compare device *and* inode to be sure they're the same file. If even you get this wrong, then surely thats a fantastic reason to expose this nicely instead of just expose the platform-specific inode (which won't even exist on windows).

bcardiff commented:
PR sent for this. I don't think we need to expose inode for file comparison in order to stay posix independent for this api.

Users are free to go to info and extract more information, but for comparison File.same? is good API.

ysbaddaden commented:
My wrong, you're both right. Let's go with `File.same?`.
